File: ui/apple/Sources/ClambhookShared/DashboardStore.swift
import Foundation

public let bandwidthSampleLimit = 60
public let maxLogLines = 200

@MainActor
public final class DashboardStore: ObservableObject {
    @Published public private(set) var status = StatusPayload()
    @Published public private(set) var profiles = ProfilesPayload()
    @Published public private(set) var servers = ServersPayload()
    @Published public private(set) var rules = RulesPayload()
    @Published public private(set) var policyGroups = PolicyGroupsPayload()
    @Published public private(set) var ruleSets = RuleSetsPayload()
    @Published public private(set) var ruleSubscriptions = RuleSubscriptionsPayload()
    @Published public private(set) var traffic = TrafficSnapshotPayload()
    @Published public private(set) var networkSettings = TunnelNetworkSettingsPayload()
    @Published public private(set) var bandwidthSamples: [BandwidthSample] = []
    @Published public private(set) var logs: [String] = []
    @Published public private(set) var apiOnline = false
    @Published public private(set) var errorText = ""
    @Published public private(set) var recoveryIssue: TunnelRecoveryIssue?

    private let api: ClambhookAPIProviding
    private let snapshotStore: FileSnapshotStore
    private var eventTask: Task<Void, Never>?
    private var logRetention: Int

    public init(api: ClambhookAPIProviding, snapshotStore: FileSnapshotStore, logRetention: Int = maxLogLines) {
        self.api = api
        self.snapshotStore = snapshotStore
        self.logRetention = min(max(logRetention, minLogRetention), maxLogRetention)
    }

    deinit {
        eventTask?.cancel()
    }

    public var activeProfile: String {
        profiles.active.isEmpty ? status.profile : profiles.active
    }

    public var currentBandwidth: BandwidthSample {
        bandwidthSamples.last ?? BandwidthSample()
    }

    public func refreshDashboard() async {
        do {
            if let dashboardAPI = api as? ClambhookDashboardProviding {
                apply(dashboard: try await dashboardAPI.dashboard())
            } else {
                let status = try await api.status()
                let profiles = try await api.profiles()
                let servers = try await api.servers()
                let policyGroups = try await api.policyGroups()
                let rules = try await api.rules()
                let traffic = try await api.traffic()
                apply(dashboard: TunnelDashboardPayload(
                    status: status,
                    profiles: profiles,
                    servers: servers,
                    rules: rules,
                    policyGroups: policyGroups,
                    ruleSets: RuleSetsPayload(profile: rules.profile, statuses: rules.ruleSets),
                    traffic: traffic
                ))
            }
            self.apiOnline = true
            updateRecoveryIssueFromTraffic()
            await persistSnapshot()
        } catch {
            self.apiOnline = false
            setRecoveryIssue(TunnelRecoveryClassifier.issue(for: error))
            await persistSnapshot()
        }
    }

    private func apply(dashboard: TunnelDashboardPayload) {
        status = dashboard.status
        profiles = dashboard.profiles
        servers = dashboard.servers
        rules = dashboard.rules
        policyGroups = dashboard.policyGroups
        ruleSets = dashboard.ruleSets
        ruleSubscriptions = dashboard.ruleSubscriptions
        traffic = dashboard.traffic
        networkSettings = dashboard.networkSettings
    }

    public func refreshStatus() async {
        do {
            status = try await api.status()
            policyGroups = try await api.policyGroups()
            traffic = try await api.traffic()
            apiOnline = true
            updateRecoveryIssueFromTraffic()
            await persistSnapshot()
        } catch {
            apiOnline = false
            setRecoveryIssue(TunnelRecoveryClassifier.issue(for: error))
            await persistSnapshot()
        }
    }

    public func connect() async {
        await performAction { try await api.connect() }
    }

    public func disconnect() async {
        await performAction { try await api.disconnect() }
    }

    public func setActiveProfile(_ name: String) async {
        guard name != activeProfile else { return }
        await performAction { try await api.setActiveProfile(name) }
    }

    public func startEventStream(from client: ClambhookAPIClient, reconnectDelay: Duration = .seconds(2)) {
        eventTask?.cancel()
        eventTask = Task { [weak self] in
            while !Task.isCancelled {
                do {
                    for try await event in client.eventStream() {
                        await self?.apply(event: event)
                    }
                } catch {
                    await MainActor.run {
                        self?.setRecoveryIssue(TunnelRecoveryClassifier.issue(forRawError: "events: \(error.localizedDescription)"))
                    }
                }
                try? await Task.sleep(for: reconnectDelay)
            }
        }
    }

    public func stopEventStream() {
        eventTask?.cancel()
        eventTask = nil
    }

    public func updateLogRetention(_ value: Int) {
        logRetention = min(max(value, minLogRetention), maxLogRetention)
        trimLogs()
    }

    public func apply(event: DaemonEvent) async {
        switch event.type {
        case "connection.bytes":
            applyConnectionBytes(event)
        case "log.line":
            applyLogLine(event)
        default:
            if event.type.hasPrefix("connection.") || event.type.hasPrefix("rule.") || event.type.hasPrefix("hop.") {
                await refreshStatus()
                return
            }
        }
        await persistSnapshot()
    }

    public func setRecoveryIssue(_ issue: TunnelRecoveryIssue?) {
        recoveryIssue = issue
        errorText = issue?.message ?? ""
    }

    public func clearRecoveryIssue(kind: TunnelRecoveryKind? = nil) {
        guard let kind else {
            setRecoveryIssue(nil)
            return
        }
        if recoveryIssue?.kind == kind {
            setRecoveryIssue(nil)
        }
    }

    private func performAction(_ action: () async throws -> Void) async {
        do {
            try await action()
            await refreshDashboard()
        } catch {
            apiOnline = false
            setRecoveryIssue(TunnelRecoveryClassifier.issue(for: error))
            await persistSnapshot()
        }
    }

    private func updateRecoveryIssueFromTraffic() {
        if let issue = trafficRecoveryIssue() {
            setRecoveryIssue(issue)
        } else {
            setRecoveryIssue(nil)
        }
    }

    private func trafficRecoveryIssue() -> TunnelRecoveryIssue? {
        traffic.connections
            .sorted { $0.updatedTsNs > $1.updatedTsNs }
            .lazy
            .flatMap(\.hops)
            .compactMap { hop -> TunnelRecoveryIssue? in
                guard !hop.error.isEmpty else { return nil }
                let issue = TunnelRecoveryClassifier.issue(forRawError: hop.error)
                return issue.kind == .generic ? nil : issue
            }
            .first
    }

    private func applyConnectionBytes(_ event: DaemonEvent) {
        guard
            let rxDelta = event.data["rx_delta"]?.doubleValue,
            let txDelta = event.data["tx_delta"]?.doubleValue,
            let intervalNs = event.data["interval_ns"]?.doubleValue,
            intervalNs > 0
        else {
            return
        }
        let seconds = intervalNs / 1_000_000_000
        let sample = BandwidthSample(rxBps: rxDelta / seconds, txBps: txDelta / seconds)
        bandwidthSamples.append(sample)
        if bandwidthSamples.count > bandwidthSampleLimit {
            bandwidthSamples.removeFirst(bandwidthSamples.count - bandwidthSampleLimit)
        }
        applyTrafficBytes(event, sample: sample, rxDelta: rxDelta, txDelta: txDelta)
    }

    private func applyTrafficBytes(_ event: DaemonEvent, sample: BandwidthSample, rxDelta: Double, txDelta: Double) {
        guard let connID = event.data["conn_id"]?.stringValue,
              let index = traffic.connections.firstIndex(where: { $0.connID == connID }) else {
            return
        }
        let oldRxBps = traffic.connections[index].rxBps
        let oldTxBps = traffic.connections[index].txBps
        traffic.connections[index].rxBps = sample.rxBps
        traffic.connections[index].txBps = sample.txBps
        traffic.connections[index].rxTotal += UInt64(rxDelta)
        traffic.connections[index].txTotal += UInt64(txDelta)
        traffic.summary.rxBps += sample.rxBps - oldRxBps
        traffic.summary.txBps += sample.txBps - oldTxBps
        traffic.summary.rxTotal += UInt64(rxDelta)
        traffic.summary.txTotal += UInt64(txDelta)
    }

    private func applyLogLine(_ event: DaemonEvent) {
        guard let line = event.data["line"]?.stringValue else {
            return
        }
        logs.append(line)
        trimLogs()
    }

    private func trimLogs() {
        if logs.count > logRetention {
            logs.removeFirst(logs.count - logRetention)
        }
    }

    private func makeSnapshot() -> DashboardSnapshot {
        let activeConnections = status.listeners.reduce(0) { $0 + $1.activeConns }
        return DashboardSnapshot(
            updatedAt: Date(),
            apiOnline: apiOnline,
            running: status.running,
            profile: activeProfile,
            listenerCount: status.listeners.count,
            activeConnections: activeConnections,
            rxBps: currentBandwidth.rxBps,
            txBps: currentBandwidth.txBps,
            logs: Array(logs.suffix(10))
        )
    }

    private func persistSnapshot() async {
        try? await snapshotStore.save(makeSnapshot())
    }
}

File: ui/linux/src/main/kotlin/com/clambhook/linux/store/DashboardStore.kt
package com.clambhook.linux.store

import com.clambhook.linux.api.ClambhookApi
import com.clambhook.linux.model.*
import com.clambhook.linux.settings.MAX_LOG_RETENTION
import com.clambhook.linux.settings.MIN_LOG_RETENTION
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

const val BANDWIDTH_SAMPLE_LIMIT = 60
const val MAX_LOG_LINES = 200

data class DashboardState(
    val status: StatusPayload = StatusPayload(),
    val profiles: ProfilesPayload = ProfilesPayload(),
    val servers: ServersPayload = ServersPayload(),
    val rules: RulesPayload = RulesPayload(),
    val traffic: TrafficSnapshotPayload = TrafficSnapshotPayload(),
    val bandwidthSamples: List<BandwidthSample> = emptyList(),
    val logs: List<String> = emptyList(),
    val apiOnline: Boolean = false,
    val errorText: String = ""
)

class DashboardStore(
    private val api: ClambhookApi,
    private var logRetention: Int = MAX_LOG_LINES
) {
    private val _state = MutableStateFlow(DashboardState())
    val state: StateFlow<DashboardState> = _state.asStateFlow()
    private val mutex = Mutex()

    /** Test helper: seed traffic connections so applyTrafficBytes can find them. */
    internal fun seedTrafficConnections(connections: List<TrafficConnectionPayload>) {
        _state.value = _state.value.copy(traffic = _state.value.traffic.copy(connections = connections))
    }

    fun activeProfile(): String =
        if (_state.value.profiles.active.isNotEmpty()) _state.value.profiles.active
        else _state.value.status.profile

    fun activeConnections(): Int = _state.value.status.listeners.sumOf { it.activeConns }

    fun currentBandwidth(): BandwidthSample = _state.value.bandwidthSamples.lastOrNull() ?: BandwidthSample()

    suspend fun refreshDashboard() = mutex.withLock {
        try {
            val status = api.status()
            val profiles = api.profiles()
            val servers = api.servers()
            val rules = api.rules()
            val traffic = api.traffic()
            _state.value = _state.value.copy(status = status, profiles = profiles, servers = servers, rules = rules, traffic = traffic, apiOnline = true, errorText = "")
        } catch (e: Exception) {
            _state.value = _state.value.copy(apiOnline = false, errorText = e.message ?: "error")
        }
    }

    suspend fun refreshStatus() = mutex.withLock {
        try {
            val status = api.status()
            val traffic = api.traffic()
            _state.value = _state.value.copy(status = status, traffic = traffic, apiOnline = true, errorText = "")
        } catch (e: Exception) {
            _state.value = _state.value.copy(apiOnline = false, errorText = e.message ?: "error")
        }
    }

    suspend fun connect() {
        try { api.connect(); refreshDashboard() }
        catch (e: Exception) { _state.value = _state.value.copy(apiOnline = false, errorText = e.message ?: "error") }
    }

    suspend fun disconnect() {
        try { api.disconnect(); refreshDashboard() }
        catch (e: Exception) { _state.value = _state.value.copy(apiOnline = false, errorText = e.message ?: "error") }
    }

    suspend fun setActiveProfile(name: String) {
        if (name == activeProfile()) return
        try { api.setActiveProfile(name); refreshDashboard() }
        catch (e: Exception) { _state.value = _state.value.copy(apiOnline = false, errorText = e.message ?: "error") }
    }

    suspend fun createRule(rule: RulePayload) {
        try { val rules = api.createRule(rule); _state.value = _state.value.copy(rules = rules); refreshDashboard() }
        catch (e: Exception) { _state.value = _state.value.copy(apiOnline = false, errorText = e.message ?: "error") }
    }

    suspend fun createRuleFromConnection(connection: TrafficConnectionPayload, rule: RulePayload) {
        try { val rules = api.createRuleFromConnection(connection, rule); _state.value = _state.value.copy(rules = rules); refreshDashboard() }
        catch (e: Exception) { _state.value = _state.value.copy(apiOnline = false, errorText = e.message ?: "error") }
    }

    suspend fun cleanupRule(suggestion: TrafficCleanupSuggestionPayload) {
        try { val rules = api.cleanupRule(suggestion); _state.value = _state.value.copy(rules = rules); refreshDashboard() }
        catch (e: Exception) { _state.value = _state.value.copy(apiOnline = false, errorText = e.message ?: "error") }
    }

    fun applyEvent(event: DaemonEvent) {
        when (event.type) {
            "connection.bytes" -> applyConnectionBytes(event)
            "log.line" -> applyLogLine(event)
            else -> {
                if (event.type.startsWith("connection.") || event.type.startsWith("rule.") || event.type.startsWith("hop.")) {
                    // trigger refresh — caller should call refreshStatus
                }
            }
        }
    }

    fun setLogRetention(value: Int) {
        logRetention = value.coerceIn(MIN_LOG_RETENTION, MAX_LOG_RETENTION)
        _state.value = _state.value.copy(logs = _state.value.logs.takeLast(logRetention))
    }

    fun setError(message: String) {
        if (message.trim().isEmpty()) return
        _state.value = _state.value.copy(errorText = message)
    }

    private fun applyConnectionBytes(event: DaemonEvent) {
        val intervalNs = event.doubleData("interval_ns")
        if (intervalNs <= 0) return
        val seconds = intervalNs / 1_000_000_000.0
        val sample = BandwidthSample(event.doubleData("rx_delta") / seconds, event.doubleData("tx_delta") / seconds)
        val samples = (_state.value.bandwidthSamples + sample).takeLast(BANDWIDTH_SAMPLE_LIMIT)
        _state.value = _state.value.copy(bandwidthSamples = samples)
        applyTrafficBytes(event, seconds)
    }

    private fun applyLogLine(event: DaemonEvent) {
        val line = event.stringData("line")
        if (line.isEmpty()) return
        val logs = (_state.value.logs + line).takeLast(logRetention)
        _state.value = _state.value.copy(logs = logs)
    }

    private fun applyTrafficBytes(event: DaemonEvent, seconds: Double) {
        val connId = event.stringData("conn_id")
        if (connId.isEmpty() || seconds <= 0) return
        val rxDelta = event.doubleData("rx_delta")
        val txDelta = event.doubleData("tx_delta")
        val rxBps = rxDelta / seconds
        val txBps = txDelta / seconds
        val connections = _state.value.traffic.connections.map { c ->
            if (c.connId != connId) c else c.copy(rxBps = rxBps, txBps = txBps,
                rxTotal = c.rxTotal + rxDelta.toULong(), txTotal = c.txTotal + txDelta.toULong())
        }
        val summary = _state.value.traffic.summary
        val updatedSummary = summary.copy(
            rxBps = summary.rxBps + rxBps,
            txBps = summary.txBps + txBps,
            rxTotal = summary.rxTotal + rxDelta.toULong(),
            txTotal = summary.txTotal + txDelta.toULong()
        )
        _state.value = _state.value.copy(traffic = _state.value.traffic.copy(connections = connections, summary = updatedSummary))
    }
}
